### PR TITLE
started compiling code

### PR DESCRIPTION
**Pull request by mohamed-barakat**
needs ToolsForHomalg v2022.09-01: https://github.com/homalg-project/homalg_project/pull/514
needs CompilerForCAP v2022.09-02: https://github.com/homalg-project/CAP_project/pull/1011

the following lines in PreSheavesPrecompiled_rest.gi is an example of how putting indices out of lists [here](https://github.com/homalg-project/CAP_project/blob/87200a71c4d1f14e45b5ebd5434f794818578f5b/CompilerForCAP/gap/LogicTemplates.gi#L293-L307) leads to an obvious regression:

                        hoisted_3_3 := List( hoisted_6_1, function ( i_4 )
                                return REM_INT( QUO_INT( i_4, hoisted_1_3 ), hoisted_2_3 );
                            end );
                        return [ List( hoisted_27_1, function ( i_4 )
                                      return hoisted_3_3[1 + hoisted_26_1[i_4]];
                                  end )[hoisted_1_2] ];

### CODE REVIEW

**Comment by zickgraf** [comment]
> Maybe that's sufficient? :D

It is rather meant to impress you :)

**Comment by zickgraf** [comment]
You can play around with the following logic templates to avoid the regression you mention above:
```
CapJitAddLogicTemplate(
    rec(
        variable_names := [ "end_pos", "func", "index" ],
        src_template := "List( [ 1 .. end_pos ], func )[index]",
        dst_template := "func( index )",
    )
);

CapJitAddLogicTemplate(
    rec(
        variable_names := [ "end_pos", "func", "index" ],
        src_template := "List( [ 0 .. end_pos ], func )[index]",
        dst_template := "func( index - 1 )",
    )
);
```
But this leads to new regressions, e.g. in SkeletalFinSets:
```diff
     ##
     AddUniversalMorphismIntoDirectProductWithGivenDirectProduct( cat,
         
 ########
 function ( cat_1, objects_1, T_1, tau_1, P_1 )
-    local hoisted_1_1, hoisted_2_1, hoisted_3_1, hoisted_4_1, deduped_5_1;
-    deduped_5_1 := Length( objects_1 );
-    hoisted_4_1 := [ 1 .. deduped_5_1 ];
+    local hoisted_1_1, hoisted_2_1, hoisted_3_1;
+    hoisted_3_1 := [ 1 .. Length( objects_1 ) ];
     hoisted_2_1 := List( objects_1, function ( x_2 )
             return Length( x_2 );
         end );
-    hoisted_3_1 := List( [ 0 .. deduped_5_1 - 1 ], function ( j_2 )
-            return Product( hoisted_2_1{[ 1 .. j_2 ]} );
-        end );
     hoisted_1_1 := List( tau_1, function ( x_2 )
             return AsList( x_2 );
         end );
     return ObjectifyMorphismWithSourceAndRangeForCAPWithAttributes( rec(
            ), cat_1, T_1, P_1, AsList, List( [ 1 .. Length( T_1 ) ], function ( i_2 )
-              return Sum( hoisted_4_1, function ( j_3 )
-                      return hoisted_1_1[j_3][i_2] * hoisted_3_1[j_3];
+              return Sum( hoisted_3_1, function ( j_3 )
+                      return hoisted_1_1[j_3][i_2] * Product( hoisted_2_1{[ 1 .. (j_3 - 1) ]} );
                   end );
           end ) );
 end
 ########
```
Now `Product( hoisted_2_1{[ 1 .. (j_3 - 1) ]} )` is computed multiple times for the same `j_3`. This problem is really subtle and I do not yet have a good grasp on it :/

**Comment by mohamed-barakat** [comment]
> You can play around with the following logic templates to avoid the regression you mention above:
> 
> ```
> CapJitAddLogicTemplate(
>     rec(
>         variable_names := [ "end_pos", "func", "index" ],
>         src_template := "List( [ 1 .. end_pos ], func )[index]",
>         dst_template := "func( index )",
>     )
> );
> 
> CapJitAddLogicTemplate(
>     rec(
>         variable_names := [ "end_pos", "func", "index" ],
>         src_template := "List( [ 0 .. end_pos ], func )[index]",
>         dst_template := "func( index - 1 )",
>     )
> );
> ```

First observations:
* The templates increase the compilation time significantly
* They have an obvious effect on the code
* The runtime is unfortunately unaltered

**Comment by mohamed-barakat** [comment]
Funnily the first rule introduces the desired speedup and avoids the regression:

> ```
> CapJitAddLogicTemplate(
>     rec(
>         variable_names := [ "end_pos", "func", "index" ],
>         src_template := "List( [ 1 .. end_pos ], func )[index]",
>         dst_template := "func( index )",
>     )
> );
> ```

It is used in PR #140.

**Comment by zickgraf** [comment]
> First observations:
> 
>     * The templates increase the compilation time significantly
> 
>     * They have an obvious effect on the code
> 
>     * The runtime is unfortunately unaltered

> Funnily the first rule introduces the desired speedup and avoids the regression:

Can you rephrase you observations using only the first rule? I expect:

* Faster or unaltered compilation time.
* Obvious effect on the code.
* Improvement of the runtime.

Note: I expect both rules to give a speedup in certain cases and a regression in other cases, so the first rule seems to be a lucky hit by chance.

**Comment by zickgraf** [comment]
>     * The templates increase the compilation time significantly

For the record and future reference: According to what we discussed yesterday, you meant "The templates *decrease* the compilation time significantly", correct?

**Comment by mohamed-barakat** [comment]
> > ```
> > * The templates increase the compilation time significantly
> > ```
> 
> For the record and future reference: According to what we discussed yesterday, you meant "The templates _decrease_ the compilation time significantly", correct?

Yes, it should be "decrease" not "increase".
